### PR TITLE
Implement IcuRatio to represent big numbers.

### LINKED ISSUE #4554: Implement a New Struct `IcuRatio` to Encapsulate `Ratio<BigInt>`
@sffc proposes the creation of a new struct, `IcuRatio`, to encapsulate `Ratio<BigInt>`. This would involve defining the struct as follows:

```rust
pub(crate) struct IcuRatio(Ratio<BigInt>);
```
Advantages
Implementing `IcuRatio` offers several benefits:

1. Custom Functions and Implementations: This approach allows us to add custom functions and implementations specific to the `Ratio` type, enhancing its functionality and usability.

1. Centralized Implementation Changes: If we need to change the implementation of Ratio in the future, having a dedicated struct means these changes can be made in one centralized location. This reduces the risk of inconsistencies and bugs across the codebase.

1. Generic Type Flexibility: With IcuRatio, we can more easily make the Ratio type generic. For instance, if there's a need to create a version of Ratio that uses f64, this structure would simplify that process.

1. Improved Code Quality and Bug Prevention: Adopting a similar style to what we've done with the RataDie type in icu_calendar (which wraps an i64) could enhance code quality. In the `icu_calendar` case, this approach has proven beneficial for catching bugs and maintaining cleaner code.

### PR DESCRIPTION
**Pull request by younies**
closes #4554

### CODE REVIEW

File: components/experimental/src/units/ratio.rs [hunk 1]
```diff
@@ -32,10 +34,20 @@ impl IcuRatio {
         self.0.numer()
     }
 
+    pub fn numerator_str(&self) -> String {
+        self.0.numer().to_string()
```
Comment (robertbastian) [inline components/experimental/src/units/ratio.rs:38]: issue: we generally don't want to use `to_string`, and with numbers especially the `to_string` implementation is not localized

File: components/experimental/src/units/ratio.rs [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+// TODO: implement AsULE for IcuRatio and use it in Data module.
+// TODO: add test cases for IcuRatio.
+/// A ratio type that uses `BigInt` as the underlying type.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct IcuRatio(Ratio<BigInt>);
+
+#[derive(Debug)]
+pub enum IcuRatioError {
+    BigIntParseError(num_bigint::ParseBigIntError),
+    InvalidRatioString,
```
Comment (robertbastian) [inline components/experimental/src/units/ratio.rs:25]: what's the difference between these? please document

File: components/experimental/src/units/ratio.rs [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    }
+}
+
+impl FromStr for IcuRatio {
+    type Err = IcuRatioError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        let mut parts = s.split('/');
+        let num_str = parts.next().unwrap_or("0");
+        let den_str = parts.next().unwrap_or("1");
```
Comment (robertbastian) [inline components/experimental/src/units/ratio.rs:171]: nit: if you unwrap later you won't have to parse "1" to `IcuRatio::one`

File: components/experimental/src/units/converter.rs [hunk 4]
```diff
@@ -89,14 +88,13 @@ impl OffsetConverter {
...
-    pub(crate) conversion_rate: Ratio<BigInt>,
+    pub(crate) conversion_rate: IcuRatio,
 }
 
 impl ProportionalConverter {
     /// Converts the given value from the input unit to the output unit.
-    pub fn convert(&self, value: &Ratio<BigInt>) -> Ratio<BigInt> {
-        value * &self.conversion_rate
+    pub fn convert(&self, value: &IcuRatio) -> IcuRatio {
+        value.clone() * self.conversion_rate.clone()
```
Comment (robertbastian) [inline components/experimental/src/units/converter.rs:98]: question: where does all this cloning come from?

File: components/experimental/src/units/ratio.rs (hunk 1 above)
Comment (sffc) [reply components/experimental/src/units/ratio.rs:38]: Suggestion: remove these two functions `numerator_str` and `denominator_str` until you need them. It seems easy enough to use `.numerator().to_string()` at the call site if you _really_ end up needing that (such as in datagen).

File: components/experimental/src/units/ratio.rs [hunk 5]
```diff
@@ -0,0 +1,180 @@
...
+    }
+
+    pub fn two() -> Self {
+        Self(Ratio::from_integer(2.into()))
+    }
+
+    pub fn to_f64(&self) -> Option<f64> {
+        self.0.to_f64()
+    }
+}
```
Comment (sffc) [inline components/experimental/src/units/ratio.rs:71]: Suggestion: please implement `apply_si_prefix` as suggested in https://github.com/unicode-org/icu4x/pull/4404#discussion_r1467064609 and `From<ConversionInfoULE> for IcuRatio` as suggested in https://github.com/unicode-org/icu4x/pull/4404#discussion_r1474970310

File: components/experimental/src/units/converter.rs (hunk 4 above)
Comment (younies) [reply components/experimental/src/units/converter.rs:98]: the IcuRatio could not implement `Copy` trait

File: components/experimental/src/units/ratio.rs (hunk 5 above)
Comment (sffc) [reply components/experimental/src/units/ratio.rs:77]: You can:

1. `impl From<&ConversionInfoULE>` 
3. `impl From<ConversionInfo>`
2. Make a constructor `pub fn from_conversion_info(&ConversionInfoULE)`

File: components/experimental/src/units/converter.rs (hunk 4 above)
Comment (sffc) [reply components/experimental/src/units/converter.rs:98]: Please avoid calling `.clone()` for short-lived values. It doesn't make sense here: you are allocating two ratios and then immediately throwing them away to make a third.

File: components/experimental/src/units/ratio.rs [hunk 6]
```diff
@@ -0,0 +1,193 @@
...
+use num_traits::Signed;
+use num_traits::{One, Pow, Zero};
+
+use super::provider::{Base, SiPrefix};
+
+// TODO: implement AsULE for IcuRatio and use it in Data module.
+// TODO: add test cases for IcuRatio.
+/// A ratio type that uses `BigInt` as the underlying type.
+#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
+pub struct IcuRatio(Ratio<BigInt>);
```
Comment (sffc) [inline components/experimental/src/units/ratio.rs:21]: Does it need to be `pub`? Can you make it `pub(crate)`?

**Review by sffc — COMMENTED** [review]
Architecture

File: components/experimental/src/units/ratio.rs (hunk 6 above)
Comment (younies) [reply components/experimental/src/units/ratio.rs:21]: We need to have it pub because we will use it in the datagen and also as an output of the converter.

File: components/experimental/src/units/ratio.rs (hunk 5 above)
Comment (younies) [reply components/experimental/src/units/ratio.rs:112]: I implement `factor_from...` and `offset_from...`

File: components/experimental/src/units/converter.rs (hunk 4 above)
Comment (younies) [reply components/experimental/src/units/converter.rs]: Done, but I am not happy to call lthe functions on value not references

https://github.com/unicode-org/icu4x/pull/4799/commits/e8755e57d56e4615f014b0d6593a1eea0284f27f

Comment (younies) [reply components/experimental/src/units/converter.rs]: I reverted, we could not remove clone because of the following error.

```
cannot move out of `self.conversion_rate` which is behind a shared reference
move occurs because `self.conversion_rate` has type `IcuRatio`, which does not implement the `Copy` traitrustc[Click for full compiler diagnostic](rust-analyzer-diagnostics-view:/diagnostic%20message%20%5B0%5D?0#file%3A%2F%2F%2FUsers%2Fyouniesmahmoud%2Fi18n%2Ficu4x%2Fcomponents%2Fexperimental%2Fsrc%2Funits%2Fconverter.rs)
IcuRatio
```

and we could not implement Copy for IcuRatio for the folloiwng error

```
the trait `std::marker::Copy` cannot be implemented for this typerustc[Click for full compiler diagnostic](rust-analyzer-diagnostics-view:/diagnostic%20message%20%5B0%5D?0#file%3A%2F%2F%2FUsers%2Fyouniesmahmoud%2Fi18n%2Ficu4x%2Fcomponents%2Fexperimental%2Fsrc%2Funits%2Fratio.rs)
ratio.rs(22, 21): this field does not implement `std::marker::Copy`
ratio.rs(22, 21): the `std::marker::Copy` impl for `Ratio<BigInt>` requires that `BigInt: std::marker::Copy`
core::marker
macro Copy
```

File: components/experimental/src/units/ratio.rs [hunk 7]
```diff
@@ -0,0 +1,228 @@
...
+    BigIntParseError(num_bigint::ParseBigIntError),
+
+    /// Represents an error when parsing a ratio from a string.
+    /// For example, the string "1/0/2" is invalid because it contains more than one '/' character.
+    InvalidRatioString,
+}
+
+impl IcuRatio {
+    /// Creates a new `IcuRatio` from the given numerator and denominator.
+    pub fn new(numerator: BigInt, denominator: BigInt) -> Self {
```
Comment (sffc) [inline components/experimental/src/units/ratio.rs:38]:
```suggestion
    pub fn from_big_ints(numerator: BigInt, denominator: BigInt) -> Self {
```

File: components/experimental/src/units/ratio.rs [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+        Self(Ratio::new(numerator, denominator))
+    }
+
+    /// Extracts the factor ratio from `ConversionInfoULE`.
+    pub fn factor_from_conversion_info(conversion_info: &ConversionInfoULE) -> Self {
```
Comment (sffc) [inline components/experimental/src/units/ratio.rs:43]: Suggestion: these methods can _probably_ be `pub(crate)`; I don't think you need them in datagen

File: components/experimental/src/units/ratio.rs [hunk 9]
```diff
@@ -0,0 +1,228 @@
...
+            ),
+        ))
+    }
+
+    /// Applys the given SI prefix to the ratio.
+    /// For example: if the ratio is 1/3 and the SI prefix is "kilo" (10^3),
+    ///              the result will be 1000/3.
+    /// Another example: if the ratio is 1/3 and the SI prefix is "kibi" ("2^10"),
+    ///                  the result will be 1024/3.
+    pub fn apply_si_prefix(&mut self, si_prefix: &SiPrefix) {
```
Comment (sffc) [inline components/experimental/src/units/ratio.rs:71]: Also probably can be `pub(crate)`

File: components/experimental/src/units/ratio.rs [hunk 10]
```diff
@@ -0,0 +1,228 @@
...
+
+    /// Returns the numerator of the ratio.
+    pub fn numerator(&self) -> &BigInt {
+        self.0.numer()
+    }
+
+    /// Returns the denominator of the ratio.
+    pub fn denominator(&self) -> &BigInt {
+        self.0.denom()
+    }
```
Comment (sffc) [inline components/experimental/src/units/ratio.rs:90]: Suggestion: since this is a public type, it would be better if we didn't publicly return `&BigInt` since then we must always be able to return a `&BigInt`. Once of the goals of IcuRatio was to make the internals private, so that we could eventually make this into a trait or change the underlying bigint library or something like that. Can you start with these being `pub(crate)` functions? If you need them public for some reason, then maybe make a `pub(crate)` version that returns a borrow and a public one named `numerator_bigint()` that returns `BigInt`.

File: components/experimental/src/units/converter.rs (hunk 4 above)
Comment (sffc) [reply components/experimental/src/units/converter.rs:98]: That's the default error when Rust thinks it needs an owned type, but you should be able to perform your arithmetic using borrowed types. You might be missing some impls. @robertbastian can help you debug this.

**Review by sffc — CHANGES_REQUESTED** [review]
Some more comments arising from the requirement that this is a public type.

File: components/experimental/src/units/converter.rs (hunk 4 above)
Comment (younies) [reply components/experimental/src/units/converter.rs:98]: Okay

File: components/experimental/src/units/ratio.rs (hunk 8 above)
Comment (younies) [reply components/experimental/src/units/ratio.rs]: sure, done

File: components/experimental/src/units/ratio.rs (hunk 9 above)
Comment (younies) [reply components/experimental/src/units/ratio.rs]: done

File: components/experimental/src/units/ratio.rs (hunk 10 above)
Comment (younies) [reply components/experimental/src/units/ratio.rs]: I removed them right now, I made all the functions that do not need public, to be pub(crate), and add what is needed when I am implementing FFI or use it in the datagen.

File: components/experimental/src/units/ratio.rs [hunk 11]
```diff
@@ -10,45 +10,77 @@ use core::{
...
+            ),
+        ))
+    }
+
+    /// Applys the given SI prefix to the ratio.
+    /// For example: if the ratio is 1/3 and the SI prefix is "kilo" (10^3),
+    ///              the result will be 1000/3.
+    /// Another example: if the ratio is 1/3 and the SI prefix is "kibi" ("2^10"),
+    ///                  the result will be 1024/3.
+    pub(crate) fn apply_si_prefix(&mut self, si_prefix: &SiPrefix) {
```
Comment (robertbastian) [inline components/experimental/src/units/ratio.rs:70]: nit: I'd prefer if this was `impl MulAssign` between `IcuRatio` and `&SiPrefix`, because this is inherently a multiplication.

Alternatively, you could have a `SiPrefix::as_icu_ratio` that does the match, and multiply with that.

File: components/experimental/src/units/ratio.rs [hunk 12]
```diff
@@ -167,14 +209,21 @@ impl FromStr for IcuRatio {
 
     fn from_str(s: &str) -> Result<Self, Self::Err> {
         let mut parts = s.split('/');
-        let num_str = parts.next().unwrap_or("0");
-        let den_str = parts.next().unwrap_or("1");
+
+        let numerator = match parts.next() {
+            Some(num_str) => BigInt::from_str(num_str).map_err(IcuRatioError::BigIntParseError)?,
```
Comment (robertbastian) [inline components/experimental/src/units/ratio.rs:214]: nit: you're now parsing the `BigInt`s even if you return `IcuRatioError::InvalidRatioString` later. Defer this.

File: components/experimental/src/units/converter_factory.rs [hunk 13]
```diff
@@ -141,28 +136,15 @@ impl ConverterFactory {
...
-            input_conversion_info.offset_num(),
-            input_conversion_info.offset_den(),
-        );
-
-        let output_offset = Self::extract_ratio_from_unaligned(
-            &output_conversion_info.offset_sign,
-            output_conversion_info.offset_num(),
-            output_conversion_info.offset_den(),
-        );
+        let input_offset = IcuRatio::offset_from_conversion_info(input_conversion_info);
```
[the remark this answers is not in the document]
Comment (younies) [reply components/experimental/src/units/converter_factory.rs]: good idea, done

File: components/experimental/src/units/ratio.rs (hunk 11 above)
Comment (younies) [reply components/experimental/src/units/ratio.rs]: done

File: components/experimental/src/units/ratio.rs (hunk 12 above)
Comment (younies) [reply components/experimental/src/units/ratio.rs:186]: in order to do this, we have two ways:

1- use unwrap_or: let numerator = parts.next().unwrap_or("0")
2- collect in a vec: ``` let parts: Vec<&str> = s.split('/').collect();

        if parts.len() > 2 {
            return Err(IcuRatioError::InvalidRatioString);
        }
```

given that the case of `"1/1/1"` is not common, so, I prefer to use the current approach.

Comment (robertbastian) [reply components/experimental/src/units/ratio.rs:186]: 3- Use `denominator`/`nominator` variables of type `Option<&str>`

Comment (younies) [reply components/experimental/src/units/ratio.rs:191]: done

Comment (younies) [reply components/experimental/src/units/ratio.rs:191]: done

File: components/experimental/src/units/converter.rs [hunk 14]
```diff
@@ -68,14 +67,14 @@ pub(crate) struct OffsetConverter {
...
+    pub(crate) offset: IcuRatio,
 }
 
 impl OffsetConverter {
     /// Converts the given value from the input unit to the output unit.
-    pub(crate) fn convert(&self, value: &Ratio<BigInt>) -> Ratio<BigInt> {
+    pub(crate) fn convert(&self, value: &IcuRatio) -> IcuRatio {
         let result = self.proportional.convert(value);
-        result + &self.offset
+        result + self.offset.clone()
```
Comment (sffc) [inline components/experimental/src/units/converter.rs:77]: Issue: unnecessary cloning

**Review by sffc — COMMENTED** [review]
This looks landable with one more clone you should delete. No need to re-request review from me; @robertbastian should give approval.

File: Cargo.lock
Comment (robertbastian) [inline Cargo.lock:1]: Why do you need `cargo update`? This should be a separate PR

**Review by robertbastian — CHANGES_REQUESTED** [review]

File: components/experimental/src/units/converter.rs (hunk 14 above)
Comment (younies) [reply components/experimental/src/units/converter.rs:77]: good catch.

I need to implement this too:

```
impl Add<&IcuRatio> for &IcuRatio {
    type Output = IcuRatio;

    fn add(self, rhs: &IcuRatio) -> IcuRatio {
        IcuRatio(&self.0 + &rhs.0)
    }
}
```

File: Cargo.lock
Comment (younies) [reply Cargo.lock:1]: reverted :)

**Review by robertbastian — APPROVED** [review]

**Review by sffc — COMMENTED** [review]
Architecture looks the same as yesterday. Please don't re-request review from me unless there's something that changed architecturally.

File: components/experimental/src/units/ratio.rs (hunk 6 above)
Comment (younies) [reply components/experimental/src/units/ratio.rs:21]: I added a TODO comment because I do not want to move the test cases for now. we can do it in another PR.
